raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java

```diff
@@ -0,0 +1,648 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+final class RaftClientTestContext {
+    private static final int FETCH_MAX_WAIT_MS = 0;
+
+    static final TopicPartition METADATA_PARTITION = new TopicPartition("metadata", 0);
+    static final int LOCAL_ID = 0;
+
+    static final int ELECTION_BACKOFF_MAX_MS = 100;
+    static final int ELECTION_TIMEOUT_MS = 10000;
+    // fetch timeout is usually larger than election timeout
+    static final int FETCH_TIMEOUT_MS = 50000;
+    static final int REQUEST_TIMEOUT_MS = 5000;
+    static final int RETRY_BACKOFF_MS = 50;
+
+    private final QuorumStateStore quorumStateStore;
+    private final Random random;
+
+    final KafkaRaftClient client;
+    final Metrics metrics;
+    final MockLog log;
+    final MockNetworkChannel channel;
+    final MockTime time;
+    final Set<Integer> voters;
+
+    public static final class Builder {
+        private final QuorumStateStore quorumStateStore = new MockQuorumStateStore();
+        private final Random random = Mockito.spy(new Random(1));
+        private final MockLog log = new MockLog(METADATA_PARTITION);
+        private final Set<Integer> voters;
+
+        Builder(Set<Integer> voters) {
+            this.voters = voters;
+        }
+
+        Builder withElectedLeader(int epoch, int leaderId) throws IOException {
+            quorumStateStore.writeElectionState(ElectionState.withElectedLeader(epoch, leaderId, voters));
+            return this;
+        }
+
+        Builder withUnknownLeader(int epoch) throws IOException {
+            quorumStateStore.writeElectionState(ElectionState.withUnknownLeader(epoch, voters));
+            return this;
+        }
+
+        Builder withVotedCandidate(int epoch, int votedId) throws IOException {
+            quorumStateStore.writeElectionState(ElectionState.withVotedCandidate(epoch, votedId, voters));
+            return this;
+        }
+
+        Builder updateRandom(Consumer<Random> consumer) {
+            consumer.accept(random);
+            return this;
+        }
+
+        Builder updateLog(Consumer<MockLog> consumer) {
+            consumer.accept(log);
+            return this;
+        }
+
+        RaftClientTestContext build() throws IOException {
+            MockTime time = new MockTime();
+            Metrics metrics = new Metrics(time);
+            MockNetworkChannel channel = new MockNetworkChannel();
+            LogContext logContext = new LogContext();
+            QuorumState quorum = new QuorumState(LOCAL_ID, voters, ELECTION_TIMEOUT_MS, FETCH_TIMEOUT_MS,
+                    quorumStateStore, time, logContext, random);
+
+            Map<Integer, InetSocketAddress> voterAddresses = voters.stream().collect(Collectors.toMap(
+                        Function.identity(),
+                        RaftClientTestContext::mockAddress
+                        ));
```

Comment:
nit: this indentation looks kind of funky

Reply:
Done.